src/micro.c

```diff
@@ -170,88 +246,145 @@ microGroup_AddEndpoint(microGroup *g, microEndpointConfig *cfg)
 {
-    microError *err = NULL;
-    microEndpoint *ep = NULL;
-    bool finalize = false;
-    microDoneHandler doneHandler = NULL;
-
-    if (m == NULL)
+    natsStatus s = NATS_OK;
+    if (nc == NULL || service == NULL)
         return micro_ErrorInvalidArg;
 
-    _lock_service(m);
-
-    if (m->stopped)
+    natsMutex_Lock(nc->servicesMu);
+    if (nc->services == NULL)
     {
-        _unlock_service(m);
-        return NULL;
+        nc->services = NATS_CALLOC(1, sizeof(microService *));
+        if (nc->services == NULL)
+            s = nats_setDefaultError(NATS_NO_MEMORY);
+    }
+    else
+    {
+        microService **tmp = NATS_REALLOC(nc->services, (nc->numServices + 1) * sizeof(microService *));
+        if (tmp == NULL)
+            s = nats_setDefaultError(NATS_NO_MEMORY);
+        else
+            nc->services = tmp;
     }
-    ep = m->first_ep;
 
-    for (; ep != NULL; ep = ep->next)
+    if (s == NATS_OK)
     {
-        if (err = micro_stop_endpoint(ep), err != NULL)
-        {
-            _unlock_service(m);
-            return microError_Wrapf(err, "failed to stop service '%s', stopping endpoint '%s'", m->cfg->Name, ep->name);
-        }
+        nc->services[nc->numServices] = service;
+        nc->numServices++;
     }
+    natsMutex_Unlock(nc->servicesMu);
 
-    finalize = (m->first_ep == NULL);
-    if (finalize)
+    return micro_ErrorFromStatus(s);
+}
+
+static void
+_detach_service_from_connection(natsConnection *nc, microService *m)
+{
+    if (nc == NULL || m == NULL)
+        return;
+
+    natsMutex_Lock(nc->servicesMu);
+    for (int i = 0; i < nc->numServices; i++)
     {
-        natsLib_stopServiceCallbacks(m);
-        m->stopped = true;
-        doneHandler = m->cfg->DoneHandler;
+        if (nc->services[i] != m)
+            continue;
+        
+        for (int j = i; j < nc->numServices - 1; j++)
+            nc->services[j] = nc->services[j + 1];
+        nc->numServices--;
+        break;
     }
+    natsMutex_Unlock(nc->servicesMu);
 
-    _unlock_service(m);
+    return;
+}
+
+static microError *
+_stop_service(microService *m, bool unsubscribe, bool release)
+{
+    microError      *err            = NULL;
+    microEndpoint   *ep             = NULL;
+    int             refs            = 0;
+    int             numEndpoints    = 0;
 
-    if (finalize)
+    if (m == NULL)
+        return micro_ErrorInvalidArg;
+
+    _lock_service(m);
+    if (!m->stopped)
     {
-        if (doneHandler != NULL)
-            doneHandler(m);
+        m->stopped = true;
 
-        // Relase the endpoint's server reference from `micro_add_endpoint`.
-        _release_service(m);
+        if (unsubscribe)
+        {
+            for (ep = m->first_ep; ep != NULL; ep = ep->next)
+            {
+                if (err = micro_stop_endpoint(ep), err != NULL)
+                {
+                    err = microError_Wrapf(err, "failed to stop service '%s', stopping endpoint '%s'", m->cfg->Name, ep->config->Name);
```

Comment:
In most of the new code in this PR you did not use `'%s'` but simply `%s`. You may want to be consistent. I think code here makes it a bit more clear).